File: src/components/ClipboardManager/utils/sensitive-data/detectors/detectPrivateKey.ts
import type { SensitiveMatch } from '../sensitiveData.types'

const PRIVATE_KEY_PATTERN = /-----BEGIN\s+(RSA|DSA|EC|OPENSSH|PGP|ENCRYPTED)?\s*PRIVATE KEY-----/g

export function detectPrivateKey(text: string): SensitiveMatch[] {
  const matches: SensitiveMatch[] = []
  PRIVATE_KEY_PATTERN.lastIndex = 0

  let match: RegExpExecArray | null
  while ((match = PRIVATE_KEY_PATTERN.exec(text)) !== null) {
    matches.push({
      type: 'private_key',
      label: 'Private Key',
      level: 'critical',
      start: match.index,
      end: match.index + match[0].length,
    })
  }

  return matches
}

File: src-tauri/src/commands/clipboard/analysis/sensitive/detect_private_key.rs
use super::super::utf16_offset::byte_to_utf16_offset;
use crate::types::SensitivityMatch;
use regex::Regex;
use std::sync::OnceLock;

/// Detects PEM private-key headers. Faithful port of the frontend
/// `detectPrivateKey` sensitive-data detector.
pub fn detect_private_key(text: &str) -> Vec<SensitivityMatch> {
    static RE: OnceLock<Regex> = OnceLock::new();
    let re = RE.get_or_init(|| {
        Regex::new(r"-----BEGIN\s+(RSA|DSA|EC|OPENSSH|PGP|ENCRYPTED)?\s*PRIVATE KEY-----")
            .expect("valid private-key regex")
    });

    let mut matches = Vec::new();
    for m in re.find_iter(text) {
        matches.push(SensitivityMatch {
            kind: "private_key".to_string(),
            label: "Private Key".to_string(),
            level: "critical".to_string(),
            start: byte_to_utf16_offset(text, m.start()),
            end: byte_to_utf16_offset(text, m.end()),
        });
    }
    matches
}
